lib/netext/resolver.go

```diff
@@ -0,0 +1,140 @@
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package netext
+
+import (
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/loadimpact/k6/lib"
+)
+
+// MultiResolver returns all IP addresses for the given host.
+type MultiResolver func(host string) ([]net.IP, error)
+
+// Resolver is an interface that returns DNS information about a given host.
+type Resolver interface {
+	LookupIP(host string) (net.IP, error)
+}
+
+type resolver struct {
+	resolve    MultiResolver
+	strategy   lib.DNSStrategy
+	rrm        *sync.Mutex
+	rand       *rand.Rand
+	roundRobin map[string]uint8
+}
+
+type cacheRecord struct {
+	ips        []net.IP
+	lastLookup time.Time
+}
+
+type cacheResolver struct {
+	resolver
+	ttl   time.Duration
+	cm    *sync.Mutex
+	cache map[string]cacheRecord
+}
+
+// NewResolver returns a new DNS resolver. If ttl is not 0, responses
+// will be cached per host for the specified period. The IP returned from
+// LookupIP() will be selected based on the given strategy.
+func NewResolver(actRes MultiResolver, ttl time.Duration, strategy lib.DNSStrategy) Resolver {
+	r := rand.New(rand.NewSource(time.Now().UnixNano())) // nolint: gosec
+	res := resolver{
+		resolve:    actRes,
+		strategy:   strategy,
+		rrm:        &sync.Mutex{},
+		rand:       r,
+		roundRobin: make(map[string]uint8),
+	}
+	if ttl == 0 {
+		return &res
+	}
+	return &cacheResolver{
+		resolver: res,
+		ttl:      ttl,
+		cm:       &sync.Mutex{},
+		cache:    make(map[string]cacheRecord),
+	}
+}
+
+// LookupIP returns a single IP resolved for host, selected by the
+// configured strategy.
+func (r *resolver) LookupIP(host string) (net.IP, error) {
+	ips, err := r.resolve(host)
+	if err != nil {
+		return nil, err
+	}
+	return r.selectOne(host, ips), nil
+}
+
+// LookupIP returns a single IP resolved for host, selected by the configured
+// strategy. Results are cached per host and will be refreshed if the last
+// lookup time exceeds the configured TTL (not the TTL returned in the DNS
+// record).
+func (r *cacheResolver) LookupIP(host string) (net.IP, error) {
+	r.cm.Lock()
+
+	var ips []net.IP
+	// TODO: Invalidate? When?
+	if d, ok := r.cache[host]; ok && time.Now().Before(d.lastLookup.Add(r.ttl)) {
+		ips = r.cache[host].ips
+	} else {
+		r.cm.Unlock() // The lookup could take some time, so unlock momentarily.
+		var err error
+		ips, err = r.resolve(host)
+		if err != nil {
+			return nil, err
+		}
+		r.cm.Lock()
+		r.cache[host] = cacheRecord{ips: ips, lastLookup: time.Now()}
+	}
```

Comment:
> This will mean that we will make only 1 lookup request for each host on each ttl expire instead of ... 5k if there 5k VUs making the same request and the dns server isn't really fast

Hhmm but there wouldn't be 5k lookups of the same host with 5k VUs, since the first lookup that returns will write to the cache, and the rest will read it from there, that's the whole point of it :D You can confirm this with Wireshark: there's only one lookup regardless of the number of VUs, since the resolver is used globally.

Unless I'm misunderstanding your proposal, leaving the `CacheRecord` locked while a lookup is in progress would pause all requests that could've just used the cache in the meantime. But agree that we need more tests and benchmarks for this, which I'm inclined to leave for when the custom resolver is added.

Reply:
Maybe your resolve is really fast ... (and maybe most resolves will be really fast and my proposal wouldn't do much). 
But in the current scenario because you unlock during the resolve other VUs could come see that there is nothing in the cache/it has expired and start doing the resolve as well.
My proposal basically moves this so, we unlock the cache manager so other VUs can come and either get another record from it or get blocked on the record's lock that is being currently resolved.
P.s. I have a mistake in the algorithm above that will lead to a dead lock, will fix it in a second

Reply:
Hhmm right, I see. Would you say this is a blocker then?